lib/db_ido_mysql/idomysqlconnection.cpp

```diff
@@ -891,9 +891,13 @@ bool IdoMysqlConnection::FieldToEscapedString(const String& key, const Value& va
 
 		*result = static_cast<long>(dbrefcol);
 	} else if (DbValue::IsTimestamp(value)) {
-		long ts = rawvalue;
+		// MySQL TIMESTAMP has the year-2038 problem, hence the upper limit below.
+		// Also, FROM_UNIXTIME(1) seems to be the minimum allowed value.
```

Comment:
"seems to be"? Where does that suspicion come from? At least my local MariaDB version does to 0 exactly what, I'd expect, namely returning 1970-01-01 00:00:00 UTC:

```
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 4
Server version: 11.5.2-MariaDB Arch Linux

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> set time_zone = '+00:00';
Query OK, 0 rows affected (0.001 sec)

MariaDB [(none)]> select from_unixtime(0);
+---------------------+
| from_unixtime(0)    |
+---------------------+
| 1970-01-01 00:00:00 |
+---------------------+
1 row in set (0.001 sec)

MariaDB [(none)]> select from_unixtime(1);
+---------------------+
| from_unixtime(1)    |
+---------------------+
| 1970-01-01 00:00:01 |
+---------------------+
1 row in set (0.001 sec)
```

Also, looks like this change would introduce a year 2038 problem [where there's not necessarily one](https://mariadb.com/kb/en/timestamp/#supported-values):

> This means that the TIMESTAMP data type can hold values between '1970-01-01 00:00:01' ([UTC](https://mariadb.com/kb/en/coordinated-universal-time/)) and '2038-01-19 03:14:07' ([UTC](https://mariadb.com/kb/en/coordinated-universal-time/)) ([MariaDB 11.3](https://mariadb.com/kb/en/what-is-mariadb-113/) and earlier, 32-bit platforms ) or '2106-02-07 06:28:15 UTC' (from [MariaDB 11.5](https://mariadb.com/kb/en/what-is-mariadb-115/), 64-bit platforms only).

```
MariaDB [(none)]> select from_unixtime(4102444800);
+---------------------------+
| from_unixtime(4102444800) |
+---------------------------+
| 2100-01-01 00:00:00       |
+---------------------------+
1 row in set (0.001 sec)
```

Reply:
But you can't insert those:

```
Server version: 9.0.1 MySQL Community Server - GPL

Copyright (c) 2000, 2024, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> create table lol (cat timestamp);
Query OK, 0 rows affected (0,02 sec)

mysql> insert into lol values(FROM_UNIXTIME(2147483647));
Query OK, 1 row affected (0,02 sec)

mysql> select * from lol;
+---------------------+
| cat                 |
+---------------------+
| 2038-01-19 03:14:07 |
+---------------------+
1 row in set (0,00 sec)

mysql> insert into lol values(FROM_UNIXTIME(2147483648));
ERROR 1292 (22007): Incorrect datetime value: '2038-01-19 03:14:08' for column 'cat' at row 1
mysql> select * from lol;
+---------------------+
| cat                 |
+---------------------+
| 2038-01-19 03:14:07 |
+---------------------+
1 row in set (0,00 sec)

mysql> insert into lol values(FROM_UNIXTIME(0));
ERROR 1292 (22007): Incorrect datetime value: '1970-01-01 00:00:00' for column 'cat' at row 1
mysql> insert into lol values(FROM_UNIXTIME(1));
Query OK, 1 row affected (0,01 sec)

mysql> select * from lol;
+---------------------+
| cat                 |
+---------------------+
| 2038-01-19 03:14:07 |
| 1970-01-01 00:00:01 |
+---------------------+
2 rows in set (0,00 sec)

mysql>
```

Reply:
Oh nice, looks like inconsistencies between MySQL and MariaDB, I can insert the year 2100 without a problem, but 0 does not work indeed:

```
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 6
Server version: 11.5.2-MariaDB Arch Linux

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [jbrost]> create table lol (cat timestamp);
Query OK, 0 rows affected (0.038 sec)

MariaDB [jbrost]> insert into lol values(FROM_UNIXTIME(2147483647));
Query OK, 1 row affected (0.007 sec)

MariaDB [jbrost]> insert into lol values(FROM_UNIXTIME(2147483648));
Query OK, 1 row affected (0.004 sec)

MariaDB [jbrost]> insert into lol values(FROM_UNIXTIME(0));
ERROR 1292 (22007): Incorrect datetime value: '1970-01-01 01:00:00' for column `jbrost`.`lol`.`cat` at row 1
MariaDB [jbrost]> insert into lol values(FROM_UNIXTIME(1));
Query OK, 1 row affected (0.011 sec)

MariaDB [jbrost]> insert into lol values(FROM_UNIXTIME(4102444800));
Query OK, 1 row affected (0.010 sec)

MariaDB [jbrost]> select * from lol;
+---------------------+
| cat                 |
+---------------------+
| 2038-01-19 04:14:07 |
| 2038-01-19 04:14:08 |
| 1970-01-01 01:00:01 |
| 2100-01-01 01:00:00 |
+---------------------+
4 rows in set (0.001 sec)
```

> ```
> mysql> insert into lol values(FROM_UNIXTIME(2147483648));
> ERROR 1292 (22007): Incorrect datetime value: '2038-01-19 03:14:08' for column 'cat' at row 1
> ```

So... looks like "MySQL's FROM_UNIXTIME() does NULLs such implicitly." isn't the entire truth and the same problem can also be triggered with MySQL, at least with some versions?

Reply:
> "MySQL's FROM_UNIXTIME() does NULLs such implicitly." isn't the entire truth and the same problem can also be triggered with MySQL

In your own interest, please don't fall from your chair:

```
mysql> delete from lol;
Query OK, 4 rows affected (0,00 sec)

mysql> insert into lol values(FROM_UNIXTIME(-1));
Query OK, 1 row affected (0,01 sec)

mysql> select * from lol;
+------+
| cat  |
+------+
| NULL |
+------+
1 row in set (0,01 sec)

mysql> insert into lol values(FROM_UNIXTIME(0));
ERROR 1292 (22007): Incorrect datetime value: '1970-01-01 00:00:00' for column 'cat' at row 1
mysql>
```

🙈